File: tools/applications/mcc172_control_panel.py
#!/usr/bin/env python3
"""
    MCC 172 Control Panel

    Purpose:
        Display the MCC 172 input RMS voltages

    Description:
        This app reads and displays the input voltages.
"""
import tkinter
from tkinter import messagebox
import tkinter.font
import math
from daqhats import hat_list, mcc172, HatIDs, OptionFlags, HatError

def calc_rms(data, channel, num_channels, num_samples_per_channel):
    """ Calculate RMS of a data buffer """
    value = 0.0
    for i in range(num_samples_per_channel):
        index = (i * num_channels) + channel
        value += (data[index] * data[index]) / num_samples_per_channel

    return math.sqrt(value)

class ControlApp:
    """ Control application class """

    # pylint: disable=too-many-instance-attributes

    def __init__(self, master): # pylint: disable=too-many-statements
        """ Initialize the class """
        self.master = master
        master.title("MCC 172 Control Panel")

        # Initialize variables
        self.device_open = False
        self.open_address = 0
        self.board = None
        self.scan_run = False
        self.sample_rate = 10240
        self.update_iepe = False
        self.update_rate_ms = 200

        # GUI Setup

        self.bold_font = tkinter.font.Font(
            family=tkinter.font.nametofont("TkDefaultFont")["family"],
            size=tkinter.font.nametofont("TkDefaultFont")["size"],
            weight="bold")

        # Create and organize frames
        self.top_frame = tkinter.LabelFrame(master, text="Select Device")
        self.top_frame.pack(expand=False, fill=tkinter.X)

        self.mid_frame = tkinter.LabelFrame(master, text="IEPE Select")
        self.mid_frame.pack(expand=False, fill=tkinter.X)

        self.bottom_frame = tkinter.LabelFrame(master, text="Analog Inputs")
        self.bottom_frame.pack(expand=True, fill=tkinter.BOTH)

        # Create widgets

        self.dev_label = tkinter.Label(self.top_frame, text="MCC 172 address:")
        self.dev_label.grid(row=0, column=0)

        self.open_button = tkinter.Button(self.top_frame, text="Open", width=6,
                                          command=self.pressed_open_button)

        # Get list of MCC 172 devices for the device list widget
        self.addr_list = self.list_devices()

        if not self.addr_list:
            self.device_lister = tkinter.Label(self.top_frame, text="None found")
            self.open_button.config(state=tkinter.DISABLED)
        else:
            self.device_variable = tkinter.StringVar(self.top_frame)
            self.device_variable.set(self.addr_list[0])
            self.device_lister = tkinter.OptionMenu(self.top_frame,
                                                    self.device_variable,
                                                    *self.addr_list)

        self.device_lister.grid(row=0, column=1)
        self.open_button.grid(row=0, column=2)

        self.iepe_checkboxes = []
        self.iepe_check_values = []
        self.checkboxes = []
        self.check_values = []
        self.channel_labels = []
        self.voltages = []
        for index in range(mcc172.info().NUM_AI_CHANNELS):
            # Checkboxes
            self.iepe_check_values.append(tkinter.IntVar())
            self.iepe_checkboxes.append(tkinter.Checkbutton(
                self.mid_frame, text="Channel {} IEPE".format(index),
                variable=self.iepe_check_values[index],
                command=lambda index=index: self.pressed_iepe_check(index)))
            self.iepe_checkboxes[index].grid(row=0, column=index)

            self.check_values.append(tkinter.IntVar())
            self.checkboxes.append(tkinter.Checkbutton(
                self.bottom_frame,
                variable=self.check_values[index],
                command=lambda index=index: self.pressed_check(index)))
            self.checkboxes[index].grid(row=index, column=0)
            self.checkboxes[index].select()
            # Labels
            self.channel_labels.append(tkinter.Label(
                self.bottom_frame,
                text="Ch {}".format(index), font=self.bold_font))
            self.channel_labels[index].grid(row=index, column=1)
            self.channel_labels[index].grid_configure(sticky="W")
            # Voltages
            self.voltages.append(tkinter.Label(
                self.bottom_frame, text="0.000",
                font=self.bold_font))
            self.voltages[index].grid(row=index, column=2)
            self.voltages[index].grid_configure(sticky="E")

            self.bottom_frame.grid_rowconfigure(index, weight=1)

        self.bottom_frame.grid_columnconfigure(1, weight=1)
        self.bottom_frame.grid_columnconfigure(2, weight=1)

        self.bottom_frame.bind("<Configure>", self.resize_text)

        # Disable widgets until a device is opened
        self.disable_controls()

        master.protocol('WM_DELETE_WINDOW', self.close) # exit cleanup

        icon = tkinter.PhotoImage(file='/usr/share/mcc/daqhats/icon.png')
        # pylint: disable=protected-access
        master.tk.call('wm', 'iconphoto', master._w, icon)

    def resize_text(self, _):
        """ Text resize event """
        height = self.bottom_frame.winfo_height()
        new_size = -max(12, int(height / 4))
        self.bold_font.configure(size=new_size)

    def pressed_check(self, index):
        """ Check button pressed event """
        if self.check_values[index].get() == 0:
            self.channel_labels[index].config(state=tkinter.DISABLED)
            self.voltages[index].config(state=tkinter.DISABLED)
        else:
            self.channel_labels[index].config(state=tkinter.NORMAL)
            self.voltages[index].config(state=tkinter.NORMAL)

    def pressed_iepe_check(self, _):
        """ IEPE check button pressed event. Must be handled when a scan
            is not running. """
        self.update_iepe = True

    def disable_controls(self):
        """ Disable controls when board not opened """
        # Enable the address selector
        self.device_lister.config(state=tkinter.NORMAL)
        # Disable the board controls
        for child in self.mid_frame.winfo_children():
            child.config(state=tkinter.DISABLED)
        for child in self.bottom_frame.winfo_children():
            child.config(state=tkinter.DISABLED)

    def enable_controls(self):
        """ Enable controls when board opened """
        # Disable the address selector
        self.device_lister.config(state=tkinter.DISABLED)
        # Enable the board controls
        for child in self.mid_frame.winfo_children():
            child.config(state=tkinter.NORMAL)
        for child in self.bottom_frame.winfo_children():
            child.config(state=tkinter.NORMAL)
        # Reset the channels to enabled
        for index in range(mcc172.info().NUM_AI_CHANNELS):
            self.check_values[index].set(1)

    def list_devices(self):
        """ List attached devices """
        self.dev_list = hat_list(HatIDs.MCC_172)
        addr_list = ["{}".format(dev.address) for dev in self.dev_list]
        return addr_list

    def open_device(self, address):
        """ Open the specified device """
        try:
            self.board = mcc172(address)

            self.scan_run = False

            # Read IEPE states
            for index in range(mcc172.info().NUM_AI_CHANNELS):
                enabled = self.board.iepe_config_read(index)
                self.iepe_check_values[index].set(enabled)
        except HatError:
            return False
        else:
            return True

    def close_device(self):
        """ Close the device """
        self.board = None

    def update_inputs(self):
        """ Periodically display the input values """
        if self.device_open:
            # Read a block of samples from the channel then calculate
            # RMS AC voltage (obtained during previous scan).
            if self.scan_run:
                num_channels = mcc172.info().NUM_AI_CHANNELS
                scan_results = self.board.a_in_scan_read(-1, 0)
                self.board.a_in_scan_cleanup()
                for channel in range(num_channels):
                    if self.check_values[channel].get() == 1:
                        value = calc_rms(scan_results.data,
                                         channel, num_channels,
                                         int(len(scan_results.data) / num_channels))

                        self.voltages[channel].config(
                            text="{:.3f}".format(value))

            if self.update_iepe:
                for index in range(mcc172.info().NUM_AI_CHANNELS):
                    if self.iepe_check_values[index].get() == 0:
                        self.board.iepe_config_write(index, 0)
                    else:
                        self.board.iepe_config_write(index, 1)
                self.update_iepe = False

            # Start a new scan
            self.board.a_in_scan_start(
                3, int(self.sample_rate * self.update_rate_ms / 2000), OptionFlags.DEFAULT)
            self.scan_run = True

            # schedule another update
            self.master.after(self.update_rate_ms, self.update_inputs)

    # Event handlers
    def pressed_open_button(self):
        """ Open button pressed event """
        if self.open_button.cget('text') == "Open":
            # Open the selected device
            address = int(self.device_variable.get())
            if self.open_device(address):
                self.device_open = True
                self.open_address = address

                self.enable_controls()

                self.open_button.config(text="Close")

                # Periodically read the inputs and update controls
                self.update_inputs()
            else:
                messagebox.showerror("Error", "Could not open device.")
        else:
            if self.device_open:
                self.close_device()
                self.device_open = False
            self.open_button.config(text="Open")
            self.disable_controls()

    def close(self):
        """ Close the app """
        self.master.destroy()

def main():
    """ App entry point """
    root = tkinter.Tk()
    _app = ControlApp(root)
    root.mainloop()

if __name__ == '__main__':
    main()

File: tools/applications/mcc172_control_panel.c
#include <daqhats/daqhats.h>
#include <gtk/gtk.h>
#include <math.h>
#include <stdbool.h>
#include <stdlib.h>

GtkWidget* main_window = NULL;
GtkWidget* device_address_combo = NULL;
GtkWidget* digital_frame = NULL;
GtkWidget* analog_frame = NULL;
GtkWidget** iepe_check_buttons = NULL;
GtkWidget** channel_check_buttons = NULL;
GtkWidget** voltage_labels = NULL;
guint update_thread;
int address = 0;
bool device_open = false;
bool first_scan_run = false;
bool update_iepe = false;

#define SAMPLE_RATE     10240UL
#define UPDATE_RATE_MS  200
#define MAX_CHANNELS    2
#define SCAN_COUNT      ((SAMPLE_RATE * UPDATE_RATE_MS) / 2000)
double scan_data[SCAN_COUNT * MAX_CHANNELS];

/// @brief Callback to end the program
/// @param wid - the widget (not used)
/// @param ptr - user data (not used)
void end_program(GtkWidget* wid, gpointer ptr)
{
   if (device_open)
   {
      g_source_remove(update_thread);
      mcc172_close(address);
   }
   gtk_main_quit();
}

/// @brief Callback for the check button toggle
/// @param wid - the check button that was clicked
/// @param ptr - user data (channel index)
void toggled_check_button(GtkToggleButton* wid, gpointer ptr)
{
   int index = GPOINTER_TO_INT(ptr);

   if (!device_open)
   {
      return;
   }
   if (gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(wid)))
   {
      // enable the channel
      gtk_widget_set_sensitive(voltage_labels[index], true);
   }
   else
   {
      // disable the channel
      gtk_widget_set_sensitive(voltage_labels[index], false);
   }
}

/// @brief Enable or disable the controls
/// @param enable - true to enable, false to disable
void enable_controls(bool enable)
{
   if (device_address_combo != NULL)
   {
      // The address combo is disabled when the other controls are enabled and vice versa.
      gtk_widget_set_sensitive(device_address_combo, !enable);
   }

   gtk_widget_set_sensitive(digital_frame, enable);
   gtk_widget_set_sensitive(analog_frame, enable);
}

/// @brief Calculate the RMS value of a buffer containing multiple channels of interleaved data
/// @param data - the data buffer
/// @param channel - the channel number
/// @param num_channels - the number of channels in the data
/// @param num_samples_per_channel - the number of samples for each channel
/// @return the rms value
double calc_rms(double* data, uint8_t channel, uint8_t num_channels, uint32_t num_samples_per_channel)
{
   double value = 0.0;
   for (uint32_t i = 0; i < num_samples_per_channel; i++)
   {
      uint32_t index = (i * num_channels) + channel;
      value += (data[index] * data[index]) / num_samples_per_channel;
   }
   return sqrt(value);
}

/// @brief Timer callback to update the voltage readings
/// @param ptr - not used
/// @return true to continue the timer, false to stop
int update_inputs(void* ptr)
{
   char text[16];

   // Read a block of samples from the channel(s) then calculate RMS AC voltage
   if (!device_open)
   {
      return false;
   }

   if (first_scan_run)
   {
      // read and process the data from the previous scan
      uint16_t status = 0;
      uint32_t samples_read_per_channel = 0;
      mcc172_a_in_scan_read(address, &status, -1, 0, scan_data, SCAN_COUNT * MAX_CHANNELS, &samples_read_per_channel);
      mcc172_a_in_scan_cleanup(address);
      for (int i = 0; i < mcc172_info()->NUM_AI_CHANNELS; i++)
      {
         if (gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(channel_check_buttons[i])))
         {
            double value = calc_rms(scan_data, i, MAX_CHANNELS, samples_read_per_channel);
            sprintf(text, "%.3f", value);
            gtk_label_set_text(GTK_LABEL(voltage_labels[i]), text);
         }
      }
   }

   if (update_iepe)
   {
      for (int i = 0; i < mcc172_info()->NUM_AI_CHANNELS; i++)
      {
         bool iepe = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(iepe_check_buttons[i]));
         mcc172_iepe_config_write(address, i, (uint8_t)iepe);
      }
      update_iepe = false;
   }

   // start a new scan
   mcc172_a_in_scan_start(address, 0x03, SCAN_COUNT, 0);
   first_scan_run = true;

   return true;
}

/// @brief Callback for open button clicked
/// @param wid - the button that was clicked
/// @param ptr - user data (not used)
void clicked_open_button(GtkWidget* wid, gpointer ptr)
{
   const char* label = gtk_button_get_label(GTK_BUTTON(wid));
   if (strcmp(label, "Open") == 0)
   {
      // Open the selected device
      if (device_address_combo != NULL)
      {
         char* text = gtk_combo_box_text_get_active_text(GTK_COMBO_BOX_TEXT(device_address_combo));
         sscanf(text, "%d", &address);
         if (mcc172_open(address) == RESULT_SUCCESS)
         {
            device_open = true;
            enable_controls(true);
            gtk_button_set_label(GTK_BUTTON(wid), "Close");

            update_inputs(NULL);
            update_thread = g_timeout_add(UPDATE_RATE_MS, update_inputs, NULL);
         }
      }
   }
   else
   {
      // Close the device
      g_source_remove(update_thread);
      mcc172_close(address);
      device_open = false;
      enable_controls(false);
      gtk_button_set_label(GTK_BUTTON(wid), "Open");
   }
}

/// @brief Handler for IEPE check button toggle
/// @param wid - the check button that was toggled
/// @param ptr - the channel index
void toggled_iepe_check(GtkWidget* wid, gpointer ptr)
{
   // IEPE can only be changed when a scan is not running, so set a flag then update in update_inputs().
   update_iepe = true;
}

/// @brief Set up the GUI 
void setup_controls(void)
{
   main_window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
   gtk_window_set_title(GTK_WINDOW(main_window), "MCC 172 Control Panel");
   g_signal_connect(main_window, "delete_event", G_CALLBACK(end_program), NULL);

   // create a grid for the frames
   GtkWidget* grid = gtk_grid_new();
   gtk_container_add(GTK_CONTAINER(main_window), grid);
   gtk_container_set_border_width(GTK_CONTAINER(main_window), 2);

   // create and organize frames
   GtkWidget* top_frame = gtk_frame_new("Select device");
   gtk_frame_set_shadow_type(GTK_FRAME(top_frame), GTK_SHADOW_OUT);
   gtk_grid_attach(GTK_GRID(grid), top_frame, 0, 0, 1, 1);

   GtkWidget* top_grid = gtk_grid_new();
   gtk_grid_set_column_spacing(GTK_GRID(top_grid), 2);
   // set spacing around the outside of the grid
   gtk_widget_set_margin_top(top_grid, 2);
   gtk_widget_set_margin_bottom(top_grid, 2);
   gtk_widget_set_margin_left(top_grid, 2);
   gtk_widget_set_margin_right(top_grid, 2);
   gtk_container_add(GTK_CONTAINER(top_frame), top_grid);

   // create the top frame widgets
   GtkWidget* lbl = gtk_label_new("MCC 172 address: ");
   gtk_widget_set_hexpand(lbl, TRUE);
   gtk_grid_attach(GTK_GRID(top_grid), lbl, 0, 0, 1, 1);
   
   // get the list of MCC 172 devices
   struct HatInfo* dev_list = NULL;
   int count = hat_list(HAT_ID_MCC_172, NULL);
   if (count > 0)
   {
      dev_list = (struct HatInfo*)malloc(count * sizeof(struct HatInfo));
      count = hat_list(HAT_ID_MCC_172, dev_list);

      device_address_combo = gtk_combo_box_text_new();
      gtk_widget_set_hexpand(device_address_combo, TRUE);
      gtk_grid_attach(GTK_GRID(top_grid), device_address_combo, 1, 0, 1, 1);
      char addr_str[4];
      for (int i = 0; i < count; i++)
      {
         sprintf(addr_str, "%d", dev_list[i].address);
         gtk_combo_box_text_append_text(GTK_COMBO_BOX_TEXT(device_address_combo), addr_str);
      }
      gtk_combo_box_set_active(GTK_COMBO_BOX(device_address_combo), 0);
      free(dev_list);
   }
   else
   {
      lbl = gtk_label_new("None found");
      gtk_widget_set_hexpand(lbl, TRUE);
      gtk_grid_attach(GTK_GRID(top_grid), lbl, 1, 0, 1, 1);
   }
   
   GtkWidget* open_button = gtk_button_new_with_label("Open");
   gtk_widget_set_hexpand(open_button, TRUE);
   gtk_grid_attach(GTK_GRID(top_grid), open_button, 2, 0, 1, 1);
   g_signal_connect(open_button, "clicked", G_CALLBACK(clicked_open_button), NULL);
   
   if (count == 0)
   {
      gtk_widget_set_sensitive(open_button, FALSE);
   }

   // Create the config frame   
   digital_frame = gtk_frame_new("Configuration");
   gtk_frame_set_shadow_type(GTK_FRAME(digital_frame), GTK_SHADOW_OUT);
   gtk_grid_attach(GTK_GRID(grid), digital_frame, 0, 1, 1, 1);

   GtkWidget* config_grid = gtk_grid_new();
   gtk_grid_set_column_spacing(GTK_GRID(config_grid), 2);
   // set spacing around the outside of the grid
   gtk_widget_set_margin_top(config_grid, 2);
   gtk_widget_set_margin_bottom(config_grid, 2);
   gtk_widget_set_margin_left(config_grid, 2);
   gtk_widget_set_margin_right(config_grid, 2);
   gtk_container_add(GTK_CONTAINER(digital_frame), config_grid);

   // create the config frame widgets
   iepe_check_buttons = malloc(mcc172_info()->NUM_AI_CHANNELS * sizeof(GtkWidget*));

   for (int i = 0; i < mcc172_info()->NUM_AI_CHANNELS; i++)
   {
      char text[24];
      sprintf(text, "Ch %d IEPE Enable", i);
      iepe_check_buttons[i] = gtk_check_button_new_with_label(text);
      gtk_widget_set_hexpand(iepe_check_buttons[i], TRUE);
      gtk_widget_set_vexpand(iepe_check_buttons[i], TRUE);
      g_signal_connect(iepe_check_buttons[i], "toggled", G_CALLBACK(toggled_iepe_check), GINT_TO_POINTER(i));
      gtk_grid_attach(GTK_GRID(config_grid), iepe_check_buttons[i], i, 0, 1, 1);
   }

   // create the analog frame / widgets
   analog_frame = gtk_frame_new("Analog Inputs");
   gtk_frame_set_shadow_type(GTK_FRAME(analog_frame), GTK_SHADOW_OUT);
   gtk_grid_attach(GTK_GRID(grid), analog_frame, 0, 2, 1, 1);
      
   // Use a grid for the widgets
   GtkWidget* analog_grid = gtk_grid_new();
   gtk_grid_set_row_spacing(GTK_GRID(analog_grid), 2);
   // set spacing around the outside of the grid
   gtk_widget_set_margin_top(analog_grid, 2);
   gtk_widget_set_margin_bottom(analog_grid, 2);
   gtk_widget_set_margin_left(analog_grid, 2);
   gtk_widget_set_margin_right(analog_grid, 2);
   gtk_container_add(GTK_CONTAINER(analog_frame), analog_grid);

   lbl = gtk_label_new(NULL);
   gtk_label_set_markup(GTK_LABEL(lbl), "<b>Channel</b>");
   gtk_widget_set_hexpand(lbl, TRUE);
   gtk_widget_set_vexpand(lbl, TRUE);
   gtk_grid_attach(GTK_GRID(analog_grid), lbl, 0, 0, 1, 1);
   lbl = gtk_label_new(NULL);
   gtk_label_set_markup(GTK_LABEL(lbl), "<b>RMS Voltage</b>");
   gtk_widget_set_hexpand(lbl, TRUE);
   gtk_widget_set_vexpand(lbl, TRUE);
   gtk_grid_attach(GTK_GRID(analog_grid), lbl, 1, 0, 1, 1);

   channel_check_buttons = malloc(mcc172_info()->NUM_AI_CHANNELS * sizeof(GtkWidget*));
   voltage_labels = malloc(mcc172_info()->NUM_AI_CHANNELS * sizeof(GtkWidget*));
   for (int index = 0; index < mcc172_info()->NUM_AI_CHANNELS; index++)
   {
      // Check buttons
      char text[8];
      sprintf(text, "Ch %d", index);
      channel_check_buttons[index] = gtk_check_button_new_with_label(text);
      gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(channel_check_buttons[index]), TRUE);
      gtk_widget_set_hexpand(channel_check_buttons[index], TRUE);
      gtk_widget_set_vexpand(channel_check_buttons[index], TRUE);
      gtk_widget_set_halign(channel_check_buttons[index], GTK_ALIGN_CENTER);
      g_signal_connect(channel_check_buttons[index], "toggled", G_CALLBACK(toggled_check_button), GINT_TO_POINTER(index));
      gtk_grid_attach(GTK_GRID(analog_grid), channel_check_buttons[index], 0, index + 1, 1, 1);

      // Voltages
      voltage_labels[index] = gtk_label_new("0.000");
      gtk_widget_set_hexpand(voltage_labels[index], TRUE);
      gtk_widget_set_vexpand(voltage_labels[index], TRUE);
      gtk_widget_set_halign(voltage_labels[index], GTK_ALIGN_CENTER);
      gtk_grid_attach(GTK_GRID(analog_grid), voltage_labels[index], 1, index + 1, 1, 1);
   }
}

/// @brief Clean up after destroying the GUI
void cleanup_controls(void)
{
   free(iepe_check_buttons);
   free(channel_check_buttons);
   free(voltage_labels);
}

/// @brief Main function
int main (int argc, char *argv[])
{
   // init the main window
   gtk_init(&argc, &argv);

   setup_controls();
   enable_controls(false);

   gtk_widget_show_all(main_window);
   gtk_main();

   cleanup_controls();
   return 0;
}
